cpp/src/arrow/flight/client.cc

```diff
@@ -1399,21 +1328,41 @@ Status FlightClient::DoPut(const FlightCallOptions& options,
                            std::unique_ptr<FlightStreamWriter>* stream,
                            std::unique_ptr<FlightMetadataReader>* reader) {
   RETURN_NOT_OK(CheckOpen());
-  return impl_->DoPut(options, descriptor, schema, stream, reader);
+  std::unique_ptr<internal::ClientDataStream> remote_stream;
+  RETURN_NOT_OK(impl_->DoPut(options, &remote_stream));
+  std::shared_ptr<internal::ClientDataStream> shared_stream = std::move(remote_stream);
+  *reader =
+      std::unique_ptr<FlightMetadataReader>(new ClientMetadataReader(shared_stream));
+  *stream = std::unique_ptr<FlightStreamWriter>(
+      new ClientStreamWriter(std::move(shared_stream), options.write_options,
+                             write_size_limit_bytes_, descriptor));
+  RETURN_NOT_OK((*stream)->Begin(schema, options.write_options));
+  return Status::OK();
 }
 
 Status FlightClient::DoExchange(const FlightCallOptions& options,
                                 const FlightDescriptor& descriptor,
                                 std::unique_ptr<FlightStreamWriter>* writer,
                                 std::unique_ptr<FlightStreamReader>* reader) {
   RETURN_NOT_OK(CheckOpen());
-  return impl_->DoExchange(options, descriptor, writer, reader);
+  std::unique_ptr<internal::ClientDataStream> remote_stream;
+  RETURN_NOT_OK(impl_->DoExchange(options, &remote_stream));
+  std::shared_ptr<internal::ClientDataStream> shared_stream = std::move(remote_stream);
+  *reader = std::unique_ptr<FlightStreamReader>(
+      new ClientStreamReader(shared_stream, options.read_options, options.stop_token));
+  auto stream_writer = std::unique_ptr<ClientStreamWriter>(
+      new ClientStreamWriter(std::move(shared_stream), options.write_options,
+                             write_size_limit_bytes_, descriptor));
+  RETURN_NOT_OK(stream_writer->Begin());
+  *writer = std::move(stream_writer);
+  return Status::OK();
 }
 
 Status FlightClient::Close() {
-  // gRPC doesn't offer an explicit shutdown
-  impl_.reset(nullptr);
-  // TODO(ARROW-15473): if we track ongoing RPCs, we can cancel them first
+  if (impl_) {
+    RETURN_NOT_OK(impl_->Close());
+    impl_.reset(nullptr);
```

Comment:
Still hangs 10 seconds on my test machine. Removing this `reset` looks fixes it.

Reply:
Hmm, oddly enough that just moves the hang to `GrpcDataTest.TestDoGetLargeBatch`. The issue is still the same: gRPC has an internal thread pool with one thread left, blocked on epoll_wait with a 10 second timeout. It's hard to tell what exactly created that thread, though…

Reply:
It's also very unreliable: I can run the test many many times before it hangs.

Reply:
I removed this reset but I think this is probably an upstream thing

Reply:
And that 10 second wait is hardcoded, so we can't even configure it: https://github.com/grpc/grpc/blob/0966536dc1720d65f714ee6207bd4a8c90b2e218/src/core/lib/iomgr/tcp_posix.cc#L464

Reply:
I can reproduce the hang easily on my test machine (an Arm server), will have a look when have time.
For now, maybe we can keep this `reset` as it looks reasonable? And we don't know if it's our code or upstream thing. 10 seconds delay is not a big issue.

Reply:
Ah, sounds good then. I'll fix it tomorrow.

For me it's a bit odd: sometimes it hangs very reliably, then once I poke at it, it stops hanging, then later it hangs reliably. I do want to minimize it and report it upstream but I think that can be done outside this PR.